django/utils/text.py

```diff
@@ -64,6 +67,76 @@ def _generator():
     return "".join(_generator())
 
 
+class DjangoHTMLParser(HTMLParser):
+    class TruncationCompleted(Exception):
+        pass
+
+    void_elements = {  # https://html.spec.whatwg.org/#void-elements
+        "area",
+        "base",
+        "br",
+        "col",
+        "embed",
+        "hr",
+        "img",
+        "input",
+        "link",
+        "meta",
+        "source",
+        "track",
+        "wbr",
+    }
+
+    def __init__(self, *, convert_charrefs=True):
+        super().__init__(convert_charrefs=convert_charrefs)
+        self.tags = deque()
+        self.output = ""
+
+    def handle_starttag(self, tag, attrs):
+        self.output += self.get_starttag_text()
+        if tag not in self.void_elements:
+            self.tags.appendleft(tag)
+
+    def handle_endtag(self, tag):
+        if tag not in self.void_elements:
+            self.output += f"</{tag}>"
+            self.tags.remove(tag)
```

Comment:
This will raise `ValueError` on input `foo</p>` - which is technically valid HTML for a paragraph... https://htmlparser.info/parser/#implied-tags

Perhaps we should `try/.../except ValuError/pass` ?

Reply:
TIL! Although I'm not sure I completely follow this as elsewhere it says that the an opening `<p>` is not omissible? https://htmlparser.info/syntax/#optional-tags

Nevertheless it is possible that closing tags are provided without an opening tag and so we should try and do something with that rather than just raising `ValueError`.